std/algorithm/searching.d

```diff
@@ -1792,27 +1796,44 @@ if (isRandomAccessRange!R1 && isBidirectionalRange!R2
     }
     // Stage 2: linear find
     size_t scout = needleLength - 1;
+outer:
     for (;;)
     {
         if (scout >= haystack.length)
-        {
-            return haystack[haystack.length .. haystack.length];
-        }
+            break;
         if (!binaryFun!pred(haystack[scout], needleBack))
         {
             ++scout;
             continue;
         }
         // Found a match with the last element in the needle
-        auto cand = haystack[scout + 1 - needleLength .. haystack.length];
-        if (startsWith!pred(cand, needle))
+        static if (isRandomAccessRange!R2)
         {
+            for (size_t j = scout + 1 - needleLength, k = 0; k < needleLength - 1; ++j, ++k)
+            {
+                if (!binaryFun!pred(haystack[j], needle[k]))
+                {
+                    scout += step;
+                    continue outer;
+                }
+            }
             // found
-            return cand;
+            return haystack[scout + 1 - needleLength .. haystack.length];
+        }
+        else
+        {
+            auto cand = haystack[scout + 1 - needleLength .. haystack.length];
+            // This intermediate creation of a slice is why the
+            // random access variant above is faster.
+            if (startsWith!pred(cand, needle))
+            {
+                // found
+                return cand;
+            }
+            scout += step;
         }
-        // Continue with the stride
-        scout += step;
     }
+    return haystack[haystack.length .. haystack.length];
```

Comment:
With multiple return statements not in explicit `else` branches, won't DMD not inline this? Is it possible to refactor this with an explicit `if`?